Update doc to state that CookieTempData is the default

Eilon opened the pull request:
In ASP.NET Core 2.0 the Cookie TempData provider is the default, not session state.

Fixes https://github.com/aspnet/Docs/issues/3767

[review URL](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state?branch=pr-en-us-4350)

dnfclas commented:
@Eilon,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

Eilon commented:
Here's the current doc: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state?tabs=aspnetcore2x

Eilon commented:
cc @marlon-tucker

guardrex commented on aspnetcore/fundamentals/app-state.md:
```diff
@@ -42,15 +42,15 @@ ASP.NET Core MVC exposes the [TempData](https://docs.microsoft.com/aspnet/core/a
 ## Cookie-based TempData provider 
 
-In ASP.NET Core 1.1 and higher, you can use the cookie-based TempData provider to store a user's TempData in a cookie. To enable the  cookie-based TempData provider, register the `CookieTempDataProvider` service in `ConfigureServices`:
+In ASP.NET Core 2.0 and higher, the cookie-based TempData provider is used by default to store a user's TempData in a cookie. In ASP.NET Core 1.0 and 1.1 the session state TempData provider is the default. To configure the TempData provider for an application, register a TempData provider implementation in `ConfigureServices`:
```
Add a comma ...

> In ASP.NET Core 1.0 and 1.1, the session state ...

marlon-tucker commented:
Hello

Is `services.AddSingleton...` the recommended way to do this or is `service.AddMvc().AddSessionStateTempDataProvider();`

The later is what the origial issue suggested, and as this doc is pretty high up on google results for state in ASP Core it's good to nudge people in the right direction.

I realise they achieve effectively the same thing.

Eilon commented:
I'll work on the feedback, thanks @marlon-tucker and @guardrex for reviewing!

guardrex commented:
@Eilon This covers https://github.com/aspnet/Docs/issues/3767?

Eilon commented:
@guardrex yes it does! I tried searching for bugs but didn't find that one. I searched for `TempData` and not `Temp Data` so I didn't find it.

Eilon commented:
cc @rynowak who logged https://github.com/aspnet/Docs/issues/3767.

guardrex commented on aspnetcore/fundamentals/app-state.md:
Use a relative link there ...

> \[Base64UrlTextEncoder](/aspnet/core/api/microsoft.aspnetcore.authentication.base64urltextencoder)

Use the contraction: "is not" :point_right: "isn't"

Drop the comma before "because" since the statement that follows is restrictive:

> ... is not compressed because compressing ...

Eilon replied:
BTW I didn't change that text at all, I just moved where the paragraph is. Do you feel these improvements are important?

guardrex replied:
lol I thought you typed that. :smile:

No, that's ok. I'll take care of it. I'll do a quick pass on this topic and clean it up later.

rynowak commented:
This needs some explanation about the tradeoffs and what we recommend. 

For instance you should use session if you need to store a lot of data or if you rely on temp data for application correctness. If you are using sessions in general then you probably want to choose it for temp data.

guardrex commented:
There's also a tracking issue for updates to the app-state topic: https://github.com/aspnet/Docs/issues/3216

@scottaddie @rynowak I could jump on #3216 after this PR goes in. The topic needs a little guidelines adherence and 1.x sample help anyway. I can get your notes in (along with anything else u want) on that update.

Rick-Anderson commented:
@rynowak  wrote:

> This needs some explanation about the tradeoffs and what we recommend.

> For instance you should use session if you need to store a lot of data or if you rely on temp data for application correctness. If you are using sessions in general then you probably want to choose it for temp data.

Can you provide the complete text for explanation/tradeoffs? We can wordsmith.

Eilon commented:
@Rick-Anderson I'll write something up and post here.

Eilon commented:
Ok I updated the text. I added more details on choosing a provider. I also fixed some ordering/formatting of the headings because they seemed a bit off.

rynowak commented on aspnetcore/fundamentals/app-state.md:
> Does the application use TempData only sparingly

You can drop the `only` from this. 

I think that point 2 should come first, that's the most obvious consideration to me.

rynowak commented on aspnetcore/fundamentals/app-state.md:
It would be nice to see it called out somewhere in plain English that **cookies have a size limit and thus cookie temp data has a size limit**

Eilon commented:
@rynowak 🆙 📅

rynowak commented:
You can update all you want, I already signed off.

Eilon commented:
@Rick-Anderson if this looks good to you, can you squash & merge? Or else I'm happy to do so.
